### PR TITLE
fix(sglang): get GLM-5.2-FP8 PD + DPA + MTP + kv-aware running on MI325X

### PR DESCRIPTION
**Pull request by llying-001**
# Description

GLM-5.2-FP8 1P1D on MI325X (gfx942) would not start on our engine image, and once it did,
long prompts came back quietly wrong. This fixes the image foundation and the four
engine/router bugs the bring-up exposed. Validated on 2× 8×MI325X, ROCm 7.2.0, sglang
v0.5.16, GLM-5.2-FP8 1P1D over mooncake RDMA.

The three image commits come first on purpose: everything below is only validated on that base.

The KV fix is not gfx942-specific. **Any** PD deployment doing chunked prefill over the
mooncake transport with overlap scheduling has been transferring pages while the forward that
writes them is still running. It never raises — prompts longer than one prefill chunk are just
partially wrong, the corruption boundary lands exactly on the chunk boundary, and the same
model/backends/chunk size in an aggregated server passes. Same symptom as
sgl-project/sglang#25583, which was auto-closed with no follow-up.

> [!IMPORTANT]
> **Running GLM-5.2 PD + DP-attention + MTP on the gfx942 image requires every leg to launch
> with `--json-model-override-args '{"index_share_for_mtp_iteration":false}'`.** That is not a
> tuning knob: it stands in for `sglang_dsa` patches 02b and 04, which are `--fuzz=0` diffs
> against v0.5.15.post1 and cannot apply to this image's v0.5.16 base. Leave IndexShare on and
> the decode leg deadlocks on the first request. Every 1P1D run behind this branch set it. The
> requirement is stated in the Dockerfile header, at the patch block, in
> `apply_sglang_dsa_patches.sh` and in both patch READMEs, so a deployment inherits it from the
> image rather than from this description.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Infra/Build change
- [ ] Code refactoring

## Changes

Image — land first:

* **`Dockerfile.sglang.gfx942` base → `lmsysorg/sglang:v0.5.16-rocm720-mi30x`.** GLM-5.2 MTP
  needs sglang #30839 plus GLM-5.2's own `GlmMoeDsaForCausalLMNextN` quant resolution. This
  image deliberately does not copy `patches/sglang/`: v0.5.16 fixes GLM-5.2's nextn quark
  exclude upstream, while that backport's anchor survives the refactor inside DeepSeek's _own_
  nextn path, so applying it here would rewrite a probe that is already correct.
* **gfx942 builds Mooncake with the HIP-transport gate**, i.e. runs the rebuild
  `Dockerfile.sglang` already did. The bundled Mooncake (upstream #2682, `01d1eb2a`) prefers
  HIP IPC over RDMA, so cross-node PD dies on the first request in `hipIpcOpenMemHandle` and
  every bring-up had to rebuild `engine.so` inside the running container first. Same script and
  diffs; only the hardcoded HIP arch changed, into `MC_GPU_ARCH` (default `gfx950`, so
  `Dockerfile.sglang` is unaffected).
* **`patches/sglang_disagg/patch_mooncake_early_send_wait_event.py`**, applied at build time by
  both engine images. sglang's `prefill.py` already records a completion event as the barrier,
  but only `mori` ever read it: `mooncake/conn.py` has no `wait_event`/`synchronize()` at all,
  and the overlap path that moves non-final chunks recorded none. The patch mirrors `mori`.
* **`sglang_dsa` patch 01 now reaches gfx942.** `main` added that set while this branch was
  out, baked only by `Dockerfile.sglang`; without patch 01 the aiter (HIP) paged-MQA branch
  sizes its logits from DP-padded rows while `lengths` carries the real count, so
  `Expected lengths.size(0) == B` fires at concurrency > 1 — ROCm-specific and not substituted
  by anything at runtime. Patch 01 is re-cut from a pinned diff into a self-locating idempotent
  script (its two edit sites are byte-identical on v0.5.15.post1 and v0.5.16; the diff is
  deleted rather than kept as a second source of truth), and
  `apply_sglang_dsa_patches.sh` grows a `DSA_PATCH_SET` arm — `full` is byte-for-byte the old
  gfx950 behaviour, `indexer` is the gfx942 arm. 02a is neither carried nor substituted there:
  not re-cut for v0.5.16, and not observed on that base, which is not the same as fixed.

Engine / router:

* **dsv4 detection** (`infera/engine/dsv4_gfx942.py`): `index_topk` alone meant DeepSeek-V4, so
  GLM-5.2 (`glm_moe_dsa`, `index_topk` 2048) silently got the dsv4 knobs. Trusts
  `model_type`/`architectures` first, falls back to `index_topk` only for a checkpoint that
  names no architecture.
* **kv-event bigram tokens** (`infera/router/kv_event/`, `rust/router/src/kv_event.rs`): under
  MTP the radix key is a bigram view, so tokens arrive as overlapping `(t[i], t[i+1])` pairs —
  Python hashed the pairs, Rust's int-only decode dropped them, and the cache view stayed empty
  so prefix hits never fired. Both sides now take the first element of each pair, rebuilding
  the flat slice the query side chunks.
* **NodePort-safe port allocation** (`infera/common/net.py`): the scan could land in
  Kubernetes' NodePort range, where the port binds and works over loopback but is
  connection-refused on the node IP we advertise. Both the per-DP-rank block and the single
  port skip that window; `INFERA_NODEPORT_RANGE` overrides it, documented in
  `manual/reference/environment.md`.

## Verification

* needle 5/9 → **9/9**, 29k depth sweep 4/9 → **9/9**, full correctness suite passes — with the
  prompt still really split into 4 chunks, so this is a barrier and not chunked prefill quietly
  turning itself off. Retires the `CHUNK=524288` + `MEM_FRAC=0.80` workaround.
* Mooncake wait-event patch script run against real sglang `v0.5.16` and `v0.5.15.post1`
  source: byte-identical to the diff it replaces on both, patched files compile, re-run is a
  no-op. On drift it exits 1 having written nothing (a half-applied handover corrupts just the
  same) and its build loop uses `set -eu`, so a drifted anchor fails the build instead of
  shipping a silently broken image.
* DSA indexer patch 01, same treatment against both real sources: reproduces the deleted diff
  **byte-for-byte** on v0.5.15.post1, applies and compiles on v0.5.16, no-op on re-run,
  `_p1v2_trim` present in the bytecode on both, and exit 1 writing nothing on a drifted anchor.
  The `full` arm was re-run end-to-end on a v0.5.15.post1 tree — all 8 bytecode markers, the
  nextn prerequisite and patch 2a's source marker still pass, so **gfx950 is unchanged**.
* Regression tests, each verified to fail without its fix: bigram `BlockStored` e2e (Python)
  and `decodes_sglang_bigram_batch_under_mtp` (Rust), NodePort avoidance for both the block and
  the single port, dsv4 detection. Full unit suite: 1183 passed.
* Not verified here: the two `docker build`s (no docker/GPU on the dev box). Only the gfx942
  Mooncake compile needs one, and that rebuild already succeeded against this exact image at
  runtime during bring-up.
* Not established: anything about `sglang_dsa` 02a on gfx942, and the gfx942 arm generally is
  FP8 / `dp8` / conc ≤ 128 / tilelang DSA backends — a different quantization, base and backend
  selection from the gfx950 validation. Tracked for a follow-up PR.

# Checklist:

- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by jiejingzhangamd** [comment]
check review and conflict, and document already re-org with model card receipt.

File: infera/common/net.py [hunk 1]
```diff
@@ -5,8 +5,49 @@
...
+def _reserved_nodeport_range() -> tuple[int, int] | None:
+    """The port window to keep out of, or None when the caller opted out.
+
+    ``INFERA_NODEPORT_RANGE="lo-hi"`` for a cluster that moved the range,
+    ``"none"`` to drop the guard entirely. An unparseable value keeps the
+    default rather than silently dropping the guard.
+    """
+    spec = os.environ.get("INFERA_NODEPORT_RANGE", _NODEPORT_RANGE_DEFAULT).strip()
+    if spec.lower() in ("", "none", "off"):
+        return None
```
[the remark this answers is not in the document]
Comment (llying-001) [reply infera/common/net.py:50]: Good catch — fixed in 11b5905. Empty/whitespace now falls through to the
default range instead of returning None; only `none`/`off` opt out. This
also makes the empty case consistent with the malformed case, which already
kept the default. Added test_empty_range_keeps_the_default covering "" and
"   ", and documented the exact opt-out values in environment.md.

File: infera/router/kv_event/client.py [hunk 2]
```diff
@@ -47,6 +47,21 @@ def _offset_endpoint(endpoint: str, rank: int) -> str:
...
+    Under EAGLE/MTP, SGLang keys its radix tree on bigrams, so a block's tokens
+    arrive as the overlapping pairs ``(t[i], t[i+1])``. The first element of each
+    pair rebuilds ``t[start:end]`` -- the same flat slice ``hash_request`` chunks
+    on the query side, and radix nodes split on page boundaries so the two
+    chunkings stay aligned. Hashing the pairs as-is builds a view that no request
+    can ever match.
+    """
+    if token_ids and isinstance(token_ids[0], (list, tuple)):
+        return [pair[0] for pair in token_ids]
+    return token_ids
```
[the remark this answers is not in the document]
Comment (llying-001) [reply infera/router/kv_event/client.py]: Partly addressed in 11b5905. A mixed list can't occur at runtime: SGLang
decides the bigram view once per engine from `is_eagle`, so a radix node is
pairs or bare ints for its whole length (see the note on
SglangBlockStored.token_ids). The annotation mismatch was real though, so
_flat_tokens now takes the schema's own list[int | tuple[int, int]] and casts
in each branch, which also fixes the flat path being typed as returning pairs.
Kept the single leading isinstance rather than a per-element check: this runs
on every BlockStored event, and a per-element test would also force a full
copy on the common non-MTP path.

**Review by jiejingzhangamd — APPROVED** [review]
